File: app/src/main/java/com/dublikunt/nclient/api/components/TagList.java
package com.dublikunt.nclient.api.components;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.dublikunt.nclient.api.enums.TagType;

import org.jetbrains.annotations.Contract;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class TagList implements Parcelable {
    public static final Creator<TagList> CREATOR = new Creator<>() {
        @NonNull
        @Contract("_ -> new")
        @Override
        public TagList createFromParcel(Parcel in) {
            return new TagList(in);
        }

        @NonNull
        @Contract(value = "_ -> new", pure = true)
        @Override
        public TagList[] newArray(int size) {
            return new TagList[size];
        }
    };
    private final Tags[] tagList = new Tags[TagType.values.length];

    protected TagList(@NonNull Parcel in) {
        this();
        ArrayList<Tag> list = new ArrayList<>();
        in.readTypedList(list, Tag.CREATOR);
        addTags(list);
    }

    public TagList() {
        for (TagType type : TagType.values) tagList[type.getId()] = new Tags();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeTypedList(getAllTagsList());
    }

    public Set<Tag> getAllTagsSet() {
        HashSet<Tag> tags = new HashSet<>();
        for (Tags t : tagList) tags.addAll(t);
        return tags;
    }

    public List<Tag> getAllTagsList() {
        List<Tag> tags = new ArrayList<>();
        for (Tags t : tagList) tags.addAll(t);
        return tags;
    }

    public int getCount(@NonNull TagType type) {
        return tagList[type.getId()].size();
    }

    public Tag getTag(@NonNull TagType type, int index) {
        return tagList[type.getId()].get(index);
    }

    public int getTotalCount() {
        int total = 0;
        for (Tags t : tagList) total += t.size();
        return total;
    }

    public void addTag(@NonNull Tag tag) {
        tagList[tag.getType().getId()].add(tag);
    }

    public void addTags(@NonNull Collection<? extends Tag> tags) {
        for (Tag t : tags) addTag(t);
    }

    public List<Tag> retrieveForType(@NonNull TagType type) {
        return tagList[type.getId()];
    }

    public int getLenght() {
        return tagList.length;
    }

    public void sort(Comparator<Tag> comparator) {
        for (Tags t : tagList) t.sort(comparator);
    }

    public boolean hasTag(@NonNull Tag tag) {
        return tagList[tag.getType().getId()].contains(tag);
    }

    public boolean hasTags(@NonNull Collection<Tag> tags) {
        for (Tag tag : tags) {
            if (!hasTag(tag)) {
                return false;
            }
        }
        return true;
    }

    public static class Tags extends ArrayList<Tag> {
        public Tags(int initialCapacity) {
            super(initialCapacity);
        }

        public Tags() {
        }

        public Tags(@NonNull Collection<? extends Tag> c) {
            super(c);
        }
    }
}

File: app/src/main/kotlin/com/dublikunt/nclient/api/comments/TagList.kt
package com.dublikunt.nclient.api.comments

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import com.dublikunt.nclient.enums.TagType

open class TagList() : Parcelable {
    private val tagList = arrayOfNulls<Tags>(TagType.values.size)

    protected constructor(`in`: Parcel) : this() {
        val list = ArrayList<Tag>()
        `in`.readTypedList(list, Tag.CREATOR)
        addTags(list)
    }

    init {
        for (type in TagType.values) tagList[type.id.toInt()] = Tags()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeTypedList(allTagsList)
    }

    val allTagsSet: Set<Tag>
        get() {
            val tags = HashSet<Tag>()
            for (t in tagList) tags.addAll(t!!)
            return tags
        }
    val allTagsList: List<Tag>
        get() {
            val tags: MutableList<Tag> = ArrayList()
            for (t in tagList) tags.addAll(t!!)
            return tags
        }

    fun getCount(type: TagType): Int {
        return tagList[type.id.toInt()]!!.size
    }

    fun getTag(type: TagType, index: Int): Tag {
        return tagList[type.id.toInt()]!![index]
    }

    val totalCount: Int
        get() {
            var total = 0
            for (t in tagList) total += t!!.size
            return total
        }

    fun addTag(tag: Tag) {
        tagList[tag.type.id.toInt()]!!.add(tag)
    }

    fun addTags(tags: Collection<Tag>) {
        for (t in tags) addTag(t)
    }

    fun retrieveForType(type: TagType): Tags? {
        return tagList[type.id.toInt()]
    }

    fun getLenght(): Int {
        return tagList.size
    }

    open fun hasTag(tag: Tag): Boolean {
        return tagList[tag.type.id.toInt()]!!.contains(tag)
    }

    open fun hasTags(tags: Collection<Tag>): Boolean {
        for (tag in tags) {
            if (!hasTag(tag)) {
                return false
            }
        }
        return true
    }

    fun sort(comparator: Comparator<Tag>) {
        for (t in tagList) t!!.sortWith(comparator)
    }

    class Tags : ArrayList<Tag> {
        constructor(initialCapacity: Int) : super(initialCapacity)
        constructor()
        constructor(c: Collection<Tag>) : super(c)
    }

    companion object {
        @JvmField
        val CREATOR: Creator<TagList> = object : Creator<TagList> {
            override fun createFromParcel(`in`: Parcel): TagList {
                return TagList(`in`)
            }

            override fun newArray(size: Int): Array<TagList?> {
                return arrayOfNulls(size)
            }
        }
    }
}
